### PR TITLE
Add script for MDS conversion of bucket of text files

### PR DESCRIPTION
**Pull request by irenedea**
Similar to the other data conversion scripts in foundry, this one is aimed at a continued pretraining API, which will accept a remote bucket containing text files and pretokenconcat and convert to mds for continued pretraining. 

https://databricks.atlassian.net/browse/GRT-2273

Manual Test

* Trained mpt-125m on the first 100 text files in a remote s3 bucket for 1 epoch (i.e. 6 batches)
* Text files were processed with 4 processes, EleutherAI/gpt-neox-20b tokenizer
* Tested with both mds shards uploaded to s3 and local mds shards.

https://gist.github.com/irenedea/7256479e61519b73bf7f231c7597610c

total tokens found when writing to mds: 3028992 (This is equivalent to 5 full batches with a 6th batch that has remaining samples) 

Without data duplication (StreamingDatatset `num_canonical_nodes=1`):
`time/token` is 3028992 at the end of training

With data duplication:
`time/token` is 3145728 at the end of training

num of tokens for 6 full batches: 256 *6 * 2048 = 3145728

### CODE REVIEW

File: scripts/data_prep/convert_text_to_mds.py
Comment (irenedea) [inline scripts/data_prep/convert_text_to_mds.py:154]: Do we want to support local input folders?

File: setup.py
Comment (irenedea) [inline setup.py:50]: @dakinggg Is it okay to upgrade the mosaicml version? We need at least 0.16 for ObjectStore.list_objects

File: scripts/data_prep/convert_text_to_mds.py
Comment (irenedea) [reply scripts/data_prep/convert_text_to_mds.py:117]: yes we do, both local input and output folders

File: scripts/data_prep/convert_text_to_mds.py
Comment (irenedea) [inline scripts/data_prep/convert_text_to_mds.py:178]: add comment to say this is a hack to get rid of a HF tokenizer warning

File: scripts/data_prep/convert_text_to_mds.py
Comment (dakinggg) [inline scripts/data_prep/convert_text_to_mds.py:221]: context managers on the opens in this function?

File: scripts/data_prep/convert_text_to_mds.py
Comment (dakinggg) [inline scripts/data_prep/convert_text_to_mds.py:329]: do we want to include all the args here? e.g. changing the max workers shouldn't affect whether we need to reprocess, right?

File: scripts/data_prep/convert_text_to_mds.py
Comment (dakinggg) [inline scripts/data_prep/convert_text_to_mds.py:77]: maybe change the name of this to differentiate more clearly from the `max_workers` arg?

File: tests/test_convert_text_to_mds.py
Comment (dakinggg) [inline tests/test_convert_text_to_mds.py:77]: any reason this actually needs to be mocked?

File: tests/test_convert_text_to_mds.py
Comment (dakinggg) [inline tests/test_convert_text_to_mds.py:138]: can we test when this is ragged too? like object names doesn't evenly divide into processes groups

**Review by dakinggg — COMMENTED** [review]
I think you're already in process of doing this, but would like a manual test that this script produces data we can train on, and expected dataset size looks right before merging.

File: scripts/data_prep/utils.py
Comment (irenedea) [reply scripts/data_prep/utils.py:114]: It checks that the rename has completed because the return is None. I took it from the mds example code, but I think we can remove it because os.rename should raise an exception if there are issues.

File: tests/test_convert_text_to_mds.py
Comment (irenedea) [reply tests/test_convert_text_to_mds.py:77]: multi processing unfortunately messes with the function call counting with mocking

File: scripts/data_prep/convert_text_to_mds.py
Comment (irenedea) [reply scripts/data_prep/convert_text_to_mds.py:77]: how about changing max_workers to max_mds_writer_workers?

File: scripts/data_prep/convert_text_to_mds.py
Comment (irenedea) [reply scripts/data_prep/convert_text_to_mds.py:205]: I arrived to that logic since a local path (according to the doc string of parse_uri) is where `backend == '' and bucket == ''`, so `is_remote_path => not (backend == '' and bucket == '') => backend != '' or bucket != ''`

File: scripts/data_prep/convert_text_to_mds.py
Comment (irenedea) [reply scripts/data_prep/convert_text_to_mds.py:329]: Good point, I suppose we need to include all the args except max_workers, reprocess?

File: tests/test_convert_text_to_mds.py
Comment (irenedea) [reply tests/test_convert_text_to_mds.py:77]: hmm, i think i might know how to get around that after taking a second look... let's see

Comment (irenedea) [reply tests/test_convert_text_to_mds.py:77]: yeah, that's going to be tricky to get working with nice mocking/call counts because memory isn't shared between processes. So we could count the calls by writing things to files instead of using the nice Mock.call_count if it's really important for us to test with full multiprocessing. wdyt?

Comment (dakinggg) [reply tests/test_convert_text_to_mds.py:77]: ah thats ok. we can rely on manual test for this.

File: scripts/data_prep/convert_text_to_mds.py
Comment (dakinggg) [reply scripts/data_prep/convert_text_to_mds.py:77]: sgtm

File: scripts/data_prep/convert_text_to_mds.py
Comment (dakinggg) [reply scripts/data_prep/convert_text_to_mds.py:205]: huh, fair enough. not sure if the upstream implementation is wrong or not...but we def need a backend and a bucket for our purposes here, so lets just check that.

File: scripts/data_prep/convert_text_to_mds.py
Comment (dakinggg) [reply scripts/data_prep/convert_text_to_mds.py:329]: I think thats right.

File: scripts/data_prep/utils.py
Comment (dakinggg) [reply scripts/data_prep/utils.py:114]: ah ok. yeah, those don't seem necessary

File: scripts/data_prep/convert_text_to_mds.py
Comment (irenedea) [reply scripts/data_prep/convert_text_to_mds.py:145]: do you mean min? for preventing to indexing out of bounds?

File: scripts/data_prep/utils.py
Comment (irenedea) [reply scripts/data_prep/utils.py:149]: That's correct. In this case, it is used to not to determine the remote download path, but to determine the output file path, which is usually going to be within a temporary folder.

I think I can actually just remove the need for the prefix all together and just use the full paths from object names

File: scripts/data_prep/convert_text_to_mds.py
Comment (dakinggg) [reply scripts/data_prep/convert_text_to_mds.py:145]: oops yeah. meant min

File: scripts/data_prep/utils.py
Comment (dakinggg) [reply scripts/data_prep/utils.py:149]: ah ok, makes sense.

File: scripts/data_prep/convert_text_to_mds.py
Comment (irenedea) [inline scripts/data_prep/convert_text_to_mds.py:281]: remove

File: scripts/data_prep/convert_text_to_mds.py
Comment (irenedea) [reply scripts/data_prep/convert_text_to_mds.py]: we could do this, but the functions require state and logger args (init and upload_file). The usage of those args is pretty minimal and close to none...should we just pass dummy objects? I don't really like the idea of relying on understanding the internal implementation to use the functions because it could break easily with a change

Comment (irenedea) [reply scripts/data_prep/convert_text_to_mds.py]: i think a better but more involved solution would be to refactor the composer code to move the logic into a class that doesn't implement LoggerDestination and the LoggerDestination version can call that class... wdyt?

We could keep this for now and then refactor in a separate PR(s)

File: scripts/data_prep/convert_text_to_mds.py
Comment (irenedea) [inline scripts/data_prep/convert_text_to_mds.py:209]:
```suggestion
        print(f'Converting to MDS format...')
        with MDSWriter(out=output_folder,
                       columns=columns,
                       max_mds_writer_workers=max_mds_writer_workers,
                       compression=compression) as out:
            for sample in tqdm(dataset):
                out.write(sample)
```

File: scripts/data_prep/convert_text_to_mds.py
Comment (irenedea) [inline scripts/data_prep/convert_text_to_mds.py:300]:
```suggestion
            list(executor.map(download_and_convert_starargs, args))
```

File: llmfoundry/utils/data_prep_utils.py
Comment (dakinggg) [inline llmfoundry/utils/data_prep_utils.py:21]: is this type ignore removable?

File: tests/test_convert_text_to_mds.py
Comment (dakinggg) [inline tests/test_convert_text_to_mds.py:160]: can we iter the whole thing and assert that we have the right amount of data?

File: scripts/data_prep/convert_text_to_mds.py
Comment (dakinggg) [inline scripts/data_prep/convert_text_to_mds.py:49]: I think this group thing is no longer correct. `--concat_tokens` is a required arg for this script, since this script doesn't support the no concat mode.

File: scripts/data_prep/convert_text_to_mds.py
Comment (dakinggg) [inline scripts/data_prep/convert_text_to_mds.py:90]:
```suggestion
        'If true, reprocess the input_folder to mds format. Otherwise, only reprocess upon changes to the input folder or dataset creation parameters.'
```

File: scripts/data_prep/convert_text_to_mds.py
Comment (dakinggg) [inline scripts/data_prep/convert_text_to_mds.py:124]: given the logging PR has gone in, switch this to use logging instead of print

File: scripts/data_prep/convert_text_to_mds.py
Comment (dakinggg) [inline scripts/data_prep/convert_text_to_mds.py:198]: same

File: scripts/data_prep/convert_text_to_mds.py
Comment (dakinggg) [inline scripts/data_prep/convert_text_to_mds.py:258]: maybe just make this a constant at the top of the file?

File: scripts/data_prep/convert_text_to_mds.py
Comment (dakinggg) [inline scripts/data_prep/convert_text_to_mds.py:278]: should we raise an error here if no files are found?

File: scripts/data_prep/convert_text_to_mds.py
Comment (dakinggg) [inline scripts/data_prep/convert_text_to_mds.py:283]: log

File: scripts/data_prep/convert_text_to_mds.py
Comment (dakinggg) [inline scripts/data_prep/convert_text_to_mds.py:334]: even though i think this is safe given the flow of this script, can we make this function non destructive? seems unnecessary

**Review by dakinggg — COMMENTED** [review]
LGTM pending a few small comments. Also I requested review from Karan to make sure the streaming stuff looks ok

File: tests/test_convert_text_to_mds.py
Comment (irenedea) [reply tests/test_convert_text_to_mds.py:160]: good idea!

File: llmfoundry/utils/data_prep_utils.py
Comment (karan6181) [inline llmfoundry/utils/data_prep_utils.py:1]: nit: should we update the license to `2023` ? Applicable to all the files in this repository. May not be part of this PR but worth addressing in a separate PR.

File: llmfoundry/utils/data_prep_utils.py
Comment (karan6181) [inline llmfoundry/utils/data_prep_utils.py:20]: Can MAC run this code ? In that case, I would recommend do a `os.name == 'posix'` check. 

macOS and Linux are both POSIX-compliant. Hence, an application written for macOS should run on Linux.

File: llmfoundry/utils/data_prep_utils.py
Comment (karan6181) [inline llmfoundry/utils/data_prep_utils.py:41]: Any reason not using `persistent_workers=True` and `pin_memory=True` ?

File: scripts/data_prep/convert_text_to_mds.py
Comment (irenedea) [reply scripts/data_prep/convert_text_to_mds.py]: should be log.info, yeah?

File: llmfoundry/utils/data_prep_utils.py
Comment (karan6181) [inline llmfoundry/utils/data_prep_utils.py:62]: This method removes the sub-directory folder, rename the shard files, and put everything in a root folder. This method would work, but I recommend a below method which actually create a new `index.json` in a root directory and the `basename` field in `index.json` points to the shard filepath of a sub-directory. Also, with this method, you won't need `with_id()` method.

```
def merge_index(root: str) -> None:
    """Create a new index.json in a root directory which points to the sub-directory shard files.

    Args:
        root (str): Root directory.
    """
    shard_id = 0
    infos = []
    for sub_dir in os.listdir(out_root):
        index_filename = os.path.join(root, sub_dir, 'index.json')
        obj = json.load(open(index_filename))
        for info in obj['shards']:
            old_basename = info['raw_data']['basename']
            new_basename = os.path.join(sub_dir, old_basename)
            info['raw_data']['basename'] = new_basename

            if info['zip_data'] is not None:
                old_basename = info['zip_data']['basename']
                new_basename = os.path.join(sub_dir, old_basename)
                info['zip_data']['basename'] = new_basename

            shard_id += 1
            infos.append(info)

    index_filename = os.path.join(root, 'index.json')
    obj = {
        'version': 2,
        'shards': infos,
    }
    text = json.dumps(obj, sort_keys=True)
    with open(index_filename, 'w') as out:
        out.write(text)
```

We will update the jupyter notebook with the above method as well. 

Please verify it by running your script end-to-end.

File: scripts/data_prep/convert_text_to_mds.py
Comment (karan6181) [inline scripts/data_prep/convert_text_to_mds.py:31]: tip: add a trailing comma at the end so that lint re-structure the code uniformly. For example
```suggestion
        'Convert text files into MDS format, optionally concatenating and tokenizing',
```

Please do the same for other argparse as well.

File: scripts/data_prep/convert_text_to_mds.py
Comment (karan6181) [inline scripts/data_prep/convert_text_to_mds.py:92]: You can do something like to ensure the code is within certain character limits.

```
help=f'If true, reprocess the input_folder to mds format. Otherwise, ' +
     f'only reprocess upon changes to the input folder or dataset creation parameters.'
```

File: scripts/data_prep/convert_text_to_mds.py
Comment (karan6181) [inline scripts/data_prep/convert_text_to_mds.py:112]: Method name is missing

File: scripts/data_prep/convert_text_to_mds.py
Comment (karan6181) [inline scripts/data_prep/convert_text_to_mds.py]: Is `print` a common thing in llm-foundry or one should use `logging.info()` ?

File: scripts/data_prep/convert_text_to_mds.py
Comment (karan6181) [inline scripts/data_prep/convert_text_to_mds.py:144]: Missing method comments. Please check other methods too!

File: scripts/data_prep/convert_text_to_mds.py
Comment (karan6181) [inline scripts/data_prep/convert_text_to_mds.py:203]: [MDSWriter()](https://docs.mosaicml.com/projects/streaming/en/stable/api_reference/generated/streaming.MDSWriter.html#streaming.MDSWriter) does not have a `max_mds_writer_workers` parameter !!!.

You should instantiate one `MDSWriter` per process and ensure each process has a unique `output_folder`. Happy to explain you offline on how it works.

File: scripts/data_prep/convert_text_to_mds.py
Comment (karan6181) [inline scripts/data_prep/convert_text_to_mds.py:282]: too long line. please split it.

File: scripts/data_prep/convert_text_to_mds.py
Comment (karan6181) [inline scripts/data_prep/convert_text_to_mds.py:181]: Please also validate that `tmp_dir` is different when running with multiple workers !! If you already did, then its fine.

File: llmfoundry/utils/data_prep_utils.py
Comment (dakinggg) [reply llmfoundry/utils/data_prep_utils.py:20]: IIRC it only worked on linux specifically...but not certain.

File: llmfoundry/utils/data_prep_utils.py
Comment (dakinggg) [reply llmfoundry/utils/data_prep_utils.py:41]: neither should have any impact here I think...the dataset is only being iterated once and no gpus involved

File: scripts/data_prep/convert_text_to_mds.py
Comment (dakinggg) [reply scripts/data_prep/convert_text_to_mds.py]: we _just_ merged a PR to foundry to do proper logging. already commented this on this pr :)

File: scripts/data_prep/convert_text_to_mds.py
Comment (dakinggg) [reply scripts/data_prep/convert_text_to_mds.py:203]: @karan6181 seems like this should have crashed. Why does mdswriter accept arbitrary kwargs?

File: llmfoundry/utils/data_prep_utils.py
Comment (dakinggg) [reply llmfoundry/utils/data_prep_utils.py:1]: ah yeah, maybe, will figure this out.

File: scripts/data_prep/convert_text_to_mds.py
Comment (dakinggg) [reply scripts/data_prep/convert_text_to_mds.py:181]: `tempfile.TemporaryDirectory()` gives a different dir every time you call it. so should be fine in multiple workers.

File: scripts/data_prep/convert_text_to_mds.py
Comment (irenedea) [reply scripts/data_prep/convert_text_to_mds.py]: Yep, I've already removed all prints instead of logging.info

File: scripts/data_prep/convert_text_to_mds.py
Comment (irenedea) [reply scripts/data_prep/convert_text_to_mds.py:181]: Yup! I have verified this.

File: scripts/data_prep/convert_text_to_mds.py
Comment (irenedea) [reply scripts/data_prep/convert_text_to_mds.py:203]: ah yes, good catch! that was from a refactor. MDSWriter takes arbitrary kwargs, so it wasn't caught at runtime. My intention was for that to be max_workers=max_mds_writer_workers. Is this correct? It seems like that might not be necessary because MDSWriter aways writes to a local output folder here and max_workers is stated in the docstring to be for remote uploads. 

> You should instantiate one MDSWriter per process and ensure each process has a unique output_folder. Happy to explain you offline on how it works.


Currently there is one MDSWriter and a unique output folder per process. In `get_task_args` we assign unique subdirectory per process.

File: llmfoundry/utils/data_prep_utils.py
Comment (irenedea) [reply llmfoundry/utils/data_prep_utils.py:20]: I'm actually going to remove this code altogether. It's from a refactor, but this refactor is now out of scope for this PR since the code no longer uses build_dataloader.

File: llmfoundry/utils/data_prep_utils.py
Comment (irenedea) [reply llmfoundry/utils/data_prep_utils.py:62]: @karan6181 Okay if I do this in a follow up refactor? I can do that immediately after getting this in, but this change requires a few changes across files to handle paths differently since originally we assume that the output directory is flat.

File: scripts/data_prep/convert_text_to_mds.py
Comment (dakinggg) [reply scripts/data_prep/convert_text_to_mds.py]: yeah info seems good

File: llmfoundry/utils/data_prep_utils.py
Comment (karan6181) [reply llmfoundry/utils/data_prep_utils.py:29]: Sure works for me. Thank You!

File: scripts/data_prep/convert_text_to_mds.py
Comment (karan6181) [reply scripts/data_prep/convert_text_to_mds.py]: > seems like this should have crashed. Why does mdswriter accept arbitrary kwargs?

MDSWriter uses kwargs `progress_bar` and `max_workers`. Looks like streaming does not do a strict validation on the incorrect kwargs. I will take a look at it. 

>Currently there is one MDSWriter and a unique output folder per process. In get_task_args we assign unique subdirectory per process.

Make sense. Thank You!

Comment (irenedea) [reply scripts/data_prep/convert_text_to_mds.py]: Removed the max_workers argument because it is only used for remote uploads and we are writing to a local dir.

**Review by karan6181 — COMMENTED** [review]
LGTM. Thank You!

**Review by dakinggg — APPROVED** [review]
LGTM! Lets start trying it out!

**Pull request merged**
